### PR TITLE
2019 08 01 rules engine

### PR DESCRIPTION
**Pull request by johncrane**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by yungezz** [comment]
Hi @sergey-shandar could you pls review this PR?

**Comment by sergey-shandar** [comment]
@johncrane could you resolve conflicts?

**Comment by akning-ms** [comment]
Hi @johncrane @sergey-shandar, what is the latest status for this PR as it is pending for a long time?
It is pending on ARM review?

File: specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json [hunk 1]
```diff
@@ -0,0 +1,2484 @@
...
+      "authorizationUrl": "https://login.microsoftonline.com/common/oauth2/authorize",
+      "flow": "implicit",
+      "description": "Azure Active Directory OAuth2 Flow",
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {
+    "/providers/Microsoft.Network/checkFrontDoorNameAvailability": {
```
Comment (ravbhatnagar) [inline specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json]: As per ARM RPC, checkNameAvailability should be modeled as pointed out below - https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#check-name-availability-requests

File: specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json [hunk 2]
```diff
@@ -0,0 +1,2484 @@
...
+          "default": {
+            "description": "Front Door error response describing why the operation failed.",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponse"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Network/checkFrontDoorNameAvailability": {
```
Comment (ravbhatnagar) [inline specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json]: same comment as above.

File: specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json [hunk 3]
```diff
@@ -0,0 +1,2484 @@
...
+              "$ref": "#/definitions/FrontDoor"
+            }
+          },
+          "201": {
+            "description": "Created. The request has been fulfilled and a new Front Door has been created.",
+            "schema": {
+              "$ref": "#/definitions/FrontDoor"
+            }
+          },
+          "202": {
```
Comment (ravbhatnagar) [inline specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json]: For async, using the 201 + non-terminal provisioning state + azure async operation header is the recommended path. Looks like here you are using the 202 + location header pattern. NRP uses 202 pattern for existing types. but for new types, please consider 201. There are advantages of using the 201 pattern.

File: specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json [hunk 4]
```diff
@@ -0,0 +1,2484 @@
...
+          "nextLinkName": "nextLink"
+        },
+        "x-ms-examples": {
+          "List Front Doors in a Resource Group": {
+            "$ref": "./examples/FrontdoorList.json"
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/frontDoors/{frontDoorName}": {
```
Comment (ravbhatnagar) [inline specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json]: patch support missing for top level resource.

File: specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json [hunk 5]
```diff
@@ -0,0 +1,2484 @@
...
+        },
+        "x-ms-examples": {
+          "Purge content from Front Door": {
+            "$ref": "./examples/FrontdoorPurgeContent.json"
+          }
+        },
+        "x-ms-long-running-operation": true
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/frontDoors/{frontDoorName}/frontendEndpoints/{frontendEndpointName}/enableHttps": {
```
Comment (ravbhatnagar) [inline specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json]: can the enable and disable https scenario be achieved by a PATCH operation on frontdoor resource itself - like updating the property on frontdoor resource which contains this setting.

File: specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json [hunk 6]
```diff
@@ -0,0 +1,2484 @@
...
+              "$ref": "#/definitions/RulesEngine"
+            }
+          },
+          "201": {
+            "description": "Created. The request has been fulfilled and a new RulesEngine has been created.",
+            "schema": {
+              "$ref": "#/definitions/RulesEngine"
+            }
+          },
+          "202": {
```
Comment (ravbhatnagar) [inline specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json]: see above comment on async. 201 recommended

File: specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json [hunk 7]
```diff
@@ -0,0 +1,2484 @@
...
+          "type": "string",
+          "x-ms-enum": {
+            "name": "CustomHttpsProvisioningSubstate",
+            "modelAsString": true
+          }
+        },
+        "customHttpsConfiguration": {
+          "description": "The configuration specifying how to enable HTTPS",
+          "readOnly": true,
+          "$ref": "#/definitions/CustomHt
```
Comment (ravbhatnagar) [inline specification/frontdoor/resource-manager/Microsoft.Network/stable/2019-08-01/frontdoor.json]: not sure if these scenarios could be achieved through Azure policy. Have you checked with Policy team?

**Review by ravbhatnagar — COMMENTED** [review]
it may be quicker to get this reviewed and signed off over a call. So please set up a review with armapireview.

**Comment by akning-ms** [comment]
Hi @johncrane, as the PR is pending for a long time,  let us know whether you still need it?

**Comment by akning-ms** [comment]
close it as no response for a long time, pls reopen it when it need

**Pull request closed without merging**